array_count, array_prioritize, array_replace added to expressions

domi4484 opened the pull request:
## Description

Added some of the missing Array functions from ArrayPlus into Core
- array_count
- array_prioritize
-  array_replace

See #33302 


<!--
  IMPORTANT NOTES FOR FIRST TIME CONTRIBUTORS
  ===========================================

  Congratulations, you are about to make a pull request to QGIS! To make this as easy and pleasurable for everyone, please take the time to read these lines before opening the pull request.

  Include a few sentences describing the overall goals for this pull request (PR). If applicable also add screenshots or - even better - screencasts.
  Include both: *what* you changed and *why* you changed it.

  If this is a pull request that adds new functionality which needs documentation, give an especially detailed explanation.
  In this case, start with a short abstract and then write some text that can be copied 1:1 to the documentation in the best case.

  Also mention if you think this PR needs to be backported. And list relevant or fixed issues.

------------------------

  Reviewing is a process done by project maintainers, mostly on a volunteer basis. We try to keep the overhead as small as possible and appreciate if you help us to do so by checking the following list.
  Feel free to ask in a comment if you have troubles with any of them.

  - Commit messages are descriptive and explain the rationale for changes.

  - Commits which fix bugs include `Fixes #11111` at the bottom of the commit message. If this is your first pull request and you forgot to do this, write the same statement into this text field with the pull request description.

  - New unit tests have been added for relevant changes

  - You have run the `scripts/prepare_commit.sh` script (https://github.com/qgis/QGIS/blob/master/.github/CONTRIBUTING.md#contributing-to-qgis) before each commit.
    If you didn't do this, you can also run `./scripts/astyle_all.sh` from your source folder.

  - You have read the QGIS Coding Standards (https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html) and this PR complies with them
-->

domi4484 committed:
array_count, array_prioritize, array_replace added to
expressions

domi4484 committed:
Fixed indentation

domi4484 committed:
Fix indentation

DelazJ commented on resources/function_help/json/array_prioritize:
```diff
@@ -0,0 +1,12 @@
+  "arguments": [
+    {"arg":"array", "description":"an array"},
+    {"arg":"array_prioritize", "description":"prioritized array" }],
+  "examples": [ { "expression":"array_prioritize(array(1, 8, 2, 5), array(5, 4, 2, 1, 3, 8))", "returns":"[ 5, 2, 1, 8 ]"}
```
The example to sort the array_minority you provide in the other thread might be interesting to show here (and/or there) also.

domi4484 replied:
Ok, should I add it now or better wait till one or the other pull request get accepted?

DelazJ reviewed (commented):
Some feedback...
Thanks @domi4484

domi4484 commented on resources/function_help/json/array_prioritize:
```diff
@@ -0,0 +1,12 @@
+  "variableLenArguments": true,
+  "arguments": [
+    {"arg":"array", "description":"an array"},
+    {"arg":"array_prioritize", "description":"prioritized array" }],
```
Maybe array_sort?

domi4484 commented on resources/function_help/json/array_prioritize:
```diff
@@ -0,0 +1,12 @@
+  "type": "function",
+  "groups": ["Arrays"],
+  "description": "Returns an array sorted by another. Missing values are added to the end.",
+  "variableLenArguments": true,
```
Thanks you are right they are only two, I'll remove this

domi4484 committed:
Apply suggestions from code review

Co-authored-by: Harrissou Sant-anna <delazj@gmail.com>

domi4484 commented on resources/function_help/json/array_replace:
```diff
@@ -0,0 +1,10 @@
+{
+  "name": "array_replace",
```
I wasn't aware of this function and I think it would make sense to extend it.
I see only one little negative point: the "replace" function is in the "String" group. Probably someone working with arrays will not search this function there.

P.S. All this could also apply to the function "length" which have his counterpart "array_length". But this could broke existing expressions... I don't know since how long is "array_length" existing
P.P.S Maybe this function should be in the group "General" instead?

DelazJ replied:
A function can be displayed in multiple groups. see e.g. length (both string and geometry)

domi4484 replied:
Oh nice than I'll do same way thanks!

DelazJ replied:
Hmmm I know I'm the one who suggested that but on a second thought, I wonder if it was a so good idea. Having 4 variants for a function leads to a huge help panel and I'm afraid it might look confusing to the end users (I forgot there was a map variant). And keeping the array_replace maybe makes the function more identifiable and inline with the other array functions? 
Others, opinion? And sorry for the extra work (though we should keep ability to replace multiple values in a row).

domi4484 replied:
I agree that it looks cleaner with separated functions. So I re-splitted the functions and added to array_replace the ability to replace by array or by map.
Don't worry for the extra work, it is still learning time for me so it is good look around a little in other corners ;)

domi4484 committed:
Expressions replace extendet to take array inputs

domi4484 committed:
resplitted replace function into replace and array_replace

array_replace extended with possipility to replace more values by arrays
or maps

nyalldawson commented on resources/function_help/json/array_count:
```diff
@@ -0,0 +1,9 @@
+  "name": "array_count",
+  "type": "function",
+  "groups": ["Arrays"],
+  "description": "Counts a given value in an array.",
```
```suggestion
  "description": "Counts the number of occurrences of a given value in an array.",
```

nyalldawson commented on resources/function_help/json/array_prioritize:
```diff
@@ -0,0 +1,12 @@
+  "name": "array_prioritize",
+  "type": "function",
+  "groups": ["Arrays"],
+  "description": "Returns an array sorted by another. Missing values are added to the end.",
```
```suggestion
  "description": "Returns an array sorted using the ordering specified in another array. Values which are present in the first array but are missing from the second array will be added to the end of the result.",
```

nyalldawson commented on resources/function_help/json/array_replace:
```diff
@@ -0,0 +1,21 @@
+  "description": "Returns an array with the supplied value, array, or map of values replaced.",
+  "variants": [
+  { "variant": "Value & array variant",
+      "variant_description": "Returns an array with the supplied value or array of values replaced by a value or an array of values.",
```
```suggestion
      "variant_description": "Returns an array with the supplied value or array of values replaced by another value or an array of values.",
```

nyalldawson commented on resources/function_help/json/array_replace:
```diff
@@ -0,0 +1,21 @@
+                    { "expression":"array_replace(array(3,2,1),array(1,2,3),array(7,8,9))", "returns":"[ 9, 8, 7 ]"},
+                    { "expression":"array_replace(array('Q','G','I','S'),array('Q','S'),'-')", "returns":"[ '-', 'G', 'I', '-' ]"} ] },
+  { "variant": "Map variant",
+      "variant_description": "Returns an array with the supplied map keys replaced by paired values.",
```
```suggestion
      "variant_description": "Returns an array with the supplied map keys replaced by their paired values.",
```

nyalldawson committed:
Apply suggestions from code review
